synapse/rest/client/models.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import TYPE_CHECKING, Optional
+
+from pydantic import BaseModel, Extra, StrictInt, StrictStr, constr, validator
+
+from synapse.util.threepids import validate_email
+
+
+class AuthenticationData(BaseModel):
+    class Config:
+        extra = Extra.allow
+
+    session: Optional[StrictStr] = None
+    type: Optional[StrictStr] = None
+
+
+class EmailRequestTokenBody(BaseModel):
+    if TYPE_CHECKING:
+        client_secret: StrictStr
+    else:
+        # See also assert_valid_client_secret()
+        client_secret: constr(
+            regex="[0-9a-zA-Z.=_-]",  # noqa: F722
+            min_length=0,
+            max_length=255,
+            strict=True,
+        )
+    email: StrictStr
+    id_access_token: Optional[StrictStr]
+    id_server: Optional[StrictStr]
+    next_link: Optional[StrictStr]
```

Comment:
There's a small footgun here. The word "optional" has two senses:
- nullable: Optional[T] might hold None, or a T
- not required: the field may be omitted, to be replaced by a default value. That value might (or might not) be None.

The senses are orthogonal, and supported by Pydantic:

```python
class M(BaseModel):
    required_not_null: StrictInt
    required_maybe_null: Optional[StrictInt]
    not_required_not_none: StrictInt = 123
    not_required_maybe_none_default_none: Optional[StrictInt] = None
    not_required_maybe_none_default_not_none: Optional[StrictInt] = 123
```

Of those five attributes, these use style 2 but I think we want style 4. Additionally, the spec says that `id_access_token` is required if `id_server` is provided. We can use a validator to enforce this.

Reply:
See https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3accountpasswordemailrequesttoken